drgn/helpers/linux/llist.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (c) 2022, Oracle and/or its affiliates.
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+"""
+Lock-less NULL terminated single linked lists
+------------
+
+The ``drgn.helpers.linux.llist`` module provides helpers for working with the
+lockless single linked list implementations (``struct llist_head`` and ``struct
+llist_node``) in :linux:`include/linux/llist.h`.
+"""
+
+from typing import Iterator, Union
+
+from drgn import NULL, Object, Type, container_of
+
+
+__all__ = (
+    "llist_empty",
+    "llist_first_entry",
+    "llist_first_entry_or_null",
+    "llist_for_each",
+    "llist_for_each_entry",
+    "llist_is_singular",
+    "llist_next_entry",
+)
+
+def llist_empty(head: Object) -> bool:
+    """
+    Return whether an llist is empty.
+
+    :param head: ``struct llist_head *``
+    """
+    return not head.first
+
+
+def llist_is_singular(head: Object) -> bool:
+    """
+    Return whether an llist has only one element.
+
+    :param head: ``struct llist_head *``
+    """
+    return head.first.value_() != 0x0 and head.first.next.value_() == 0x0
```

Comment:
A couple of things here:

1. It'd be better to read `head.first` once, both because reading the live kernel's memory from `/proc/kcore` is surprisingly slow and because it's slightly less racy.
2. It is (very slightly) more efficient to use the `bool` and `not` operators.

```suggestion
    first = head.first.read_()
    return bool(first) and not first.next
```

Reply:
Done